server/src/server.ts

```diff
@@ -0,0 +1,264 @@
+    }
+});
+
+// Turn off validation of closed documents.
+documents.onDidClose(change => {
+    if (currentSettings.Validation.enable) {
+        unscheduleValidation(change.document);
+    }
+});
+
+/**
+ * Schedule a document for glTF validation after the debounce timeout.
+ * @param textDocument The document to schedule validator for.
+ */
+function scheduleValidation(textDocument: TextDocument): void {
+    if (isLocalGltf(textDocument)) {
+        console.log('schedule ' + textDocument.uri);
+        if (documentsToValidate.indexOf(textDocument) < 0) {
+            documentsToValidate.push(textDocument);
+        }
+        if (debounceValidateTimer) {
+            clearTimeout(debounceValidateTimer);
+            debounceValidateTimer = undefined;
+        }
+        debounceValidateTimer = setTimeout(() => {
+            documentsToValidate.forEach(validateTextDocument);
+            documentsToValidate = [];
+        }, currentSettings.Validation.debounce);
+    }
+}
+
+function unscheduleValidation(textDocument: TextDocument): void {
+    var index = documentsToValidate.indexOf(textDocument);
+    if (index >= 0) {
+        console.log('un-schedule ' + textDocument.uri);
+        documentsToValidate.splice(index, 1);
+    }
+}
+
+function clearValidationTextDocument(textDocument: TextDocument): void {
+    if (isLocalGltf(textDocument)) {
+        console.log('disable validation ' + textDocument.uri);
+        const diagnostics: Diagnostic[] = [];
+        connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
+    }
+}
+
+const _driveLetterPath = /^\/[a-zA-Z]:/;
+
+// From: https://github.com/Microsoft/vscode/blob/f2c2d9c65880e44b588a0e1916423b691fff01d3/src/vs/base/common/uri.ts#L357-L374
+function getFsPath(uri) {
+    const url = Url.parse(uri);
+    const urlPath = decodeURIComponent(url.path);
+    let value: string;
+    if (url.auth && urlPath && url.protocol === 'file:') {
+        // unc path: file://shares/c$/far/boo
+        value = `//${url.auth}${urlPath}`;
+    } else if (_driveLetterPath.test(urlPath)) {
+        // windows drive letter: file:///c:/far/boo
+        value = urlPath[1].toLowerCase() + urlPath.substr(2);
+    } else {
+        // other path
+        value = urlPath;
+    }
+    return path.normalize(value);
+}
+
+function validateTextDocument(textDocument: TextDocument): void {
+    console.log('validate ' + textDocument.uri);
+
+    const fileName = getFsPath(textDocument.uri);
+    const baseName = path.basename(fileName);
+
+    const gltfText = textDocument.getText();
+    const folderName = path.resolve(fileName, '..');
+
+    const map = tryGetJsonMap(textDocument);
+    if (!map) {
+        let diagnostics: Diagnostic[] = [getDiagnostic({ message: 'Error parsing JSON document.' }, {})];
+        connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
+        return;
+    }
+
+    gltfValidator.validateString(baseName, gltfText, (uri) =>
```

Comment:
Using [custom promisified function](https://nodejs.org/api/util.html#util_custom_promisified_functions) would allow complete use of async/await throughout.

Reply:
Still working on this.  `import { promisify } from 'util';` is not working yet for me...

Reply:
Seems like my typescript definitions for `util` are newer than the backing implementation.  Typescript is aware of util.promisify and all its function signatures, but the actual `util` object at runtime is completely missing the implementation for it.  Maybe I need to bump a version number somewhere, or maybe this is directly tied to the version of Node powering Electron underneath VSCode, not sure.  I do have Node 8.9.0 installed locally, FWIW.

Reply:
What does prevent using `async` / `await` in current code? Maybe I could adjust validator's API?

Reply:
Ah, according to `https://github.com/Microsoft/vscode/issues/11660`, Node does come from Electron itself, and the version number is visible in VSCode's Help->About... menu.  The current Node version is 7.9.0.  util.promisify was introduced in version 8.0.0.  Not much we can do here.

Reply:
@lexaknyazev I think the sticking point is Node's use of a callback instead of a Promise in `readFile`.  We're essentially doing a manual promisify by wrapping it the way we do.  This doesn't actually prevent async/await on the outer call to `validateString`, in fact I use async/await in the file `validationProvider.ts` (when the user manually calls the validator).  I ended up not using it here, just because this one isn't wrapped in try/catch.  I'm going to leave this one alone for now, it's not hurting anything the way it is.